pulp_python/app/models.py

```diff
@@ -92,9 +111,128 @@ class PythonImporter(Importer):
     """
 
     TYPE = 'python'
+    project_names = ArrayField(models.TextField())
 
-    def sync(self):
+    def _fetch_inventory(self):
         """
-        Synchronize the repository with the remote repository.
+        Fetch existing content in the repository.
+        Returns:
+            set: of content filenames.
         """
-        raise NotImplementedError
+        inventory = set()
+
+        q_set = PythonPackageContent.objects.filter(repositories=self.repository)
+        q_set = q_set.only("filename")
+        for content in (c.cast() for c in q_set):
+            inventory.add(content.filename)
+        return inventory
+
+    def _fetch_remote(self):
+        remote = []
+
+        metadata_urls = [urljoin(self.feed_url, 'pypi/%s/json' % project)
+                         for project in self.project_names]
+
+
+        for metadata_url in metadata_urls:
+            parsed_url = urlparse(metadata_url)
+
+            download = self.get_futures_downloader(metadata_url, os.path.basename(parsed_url.path))
+            #download_result = download.fetch()
+            download()
+
+            metadata = json.load(open(download.writer.path))
+            for version, packages in metadata['releases'].items():
+                for package in packages:
+                    remote.append(self._parse_metadata(metadata['info'], version, package))
+
+        return remote
+
+    def _find_delta(self, inventory, remote, mirror=True):
+        remote_keys = set([content['filename'] for content in remote])
+
+        additions_set = remote_keys - inventory
+        additions = [content for content in remote if content['filename'] in additions_set]
+        if mirror:
+            removals_set = inventory - remote_keys
+            removals = [content for content in remote if content['filename'] in removals_set]
+        else:
+            removals = []
+
+        return Delta(additions=additions, removals=removals)
+
+    def _parse_metadata(cls, project, version, distribution):
+
+        package = {}
+
+        package['filename'] = distribution['filename']
+        package['packagetype'] = distribution['packagetype']
+        package['name'] = project['name']
+        package['version'] = version
+        package['metadata_version'] = '3.0'
+        package['summary'] = project.get('summary')
```

Comment:
For optional string fields (iff you make them blank=True), they should probably be,

    field = whatever.get('field') or ''